### PR TITLE
Unexpected trnas data

### PR DESCRIPTION
**Pull request by buchslava**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -20,33 +21,103 @@ function walkNonDataPointIssue(context, onIssue) {
   });
 }
 
+function dataPointProcessorTemplate(context, dataPointFileDescriptor, resultHandler) {
```
Comment (rychkog) [inline index.js]: createDatapointProcessor

File: index.js [hunk 2]
```diff
@@ -20,33 +21,103 @@ function walkNonDataPointIssue(context, onIssue) {
...
+
+            resultHandler(result);
+          });
+      },
+      onDataPointReady
+    );
+  };
+}
+
+function dataPointTransProcessorTemplate(context, dataPointFileTransDescriptor, resultHandler) {
```
Comment (rychkog) [inline index.js]: createDatapointTranslationProcessor

File: index.js [hunk 3]
```diff
@@ -20,33 +21,103 @@ function walkNonDataPointIssue(context, onIssue) {
...
+
+            resultHandler(result);
+          });
+      },
+      onDataPointReady
+    );
+  };
+}
+
+function dataPointTransByRuleProcessorTemplate(context, dataPointFileTransDescriptor, resultHandler) {
```
Comment (rychkog) [inline index.js]: createDatapointTranslationByRuleProcessor

File: index.js [hunk 4]
```diff
@@ -20,33 +21,103 @@ function walkNonDataPointIssue(context, onIssue) {
...
+  const ddfDataSet = context.ddfDataSet;
+
+  return onDataPointReady => {
+    context.ddfDataSet.getDataPoint().loadFile(
+      dataPointFileTransDescriptor,
+      (transRecord, line) => {
+        const result =
```
Comment (rychkog) [inline index.js]: what is result?

File: index.js [hunk 5]
```diff
@@ -20,33 +21,103 @@ function walkNonDataPointIssue(context, onIssue) {
...
+  const ddfDataSet = context.ddfDataSet;
+
+  return onDataPointReady => {
+    context.ddfDataSet.getDataPoint().loadFile(
+      dataPointFileTransDescriptor,
+      (transRecord, line) => {
+        Object.getOwnPropertySymbols(ddfDataPointTransRules)
+          .filter(key => context.issuesFilter.isAllowed(key))
+          .forEach(key => {
+            const result =
```
Comment (rychkog) [inline index.js]: result?

File: index.js [hunk 6]
```diff
@@ -20,33 +21,103 @@ function walkNonDataPointIssue(context, onIssue) {
...
+  const ddfDataSet = context.ddfDataSet;
+
+  return onDataPointReady => {
+    context.ddfDataSet.getDataPoint().loadFile(
+      dataPointFileDescriptor,
+      (dataPointRecord, line) => {
+        Object.getOwnPropertySymbols(ddfDataPointRules)
+          .filter(key => context.issuesFilter.isAllowed(key))
+          .forEach(key => {
+            const result = ddfDataPointRules[key]({ddfDataSet, dataPointFileDescriptor, dataPointRecord, line});
```
Comment (rychkog) [inline index.js]: result?

File: lib/data/directory-descriptor.js [hunk 7]
```diff
@@ -16,6 +17,35 @@ class DirectoryDescriptor {
     this.errors = [];
   }
 
+  getTranslations(onTranslationsReady) {
+    if (_.isEmpty(this.dataPackage.getTranslations()) || !_.isArray(this.dataPackage.getTranslations())) {
```
Comment (rychkog) [inline lib/data/directory-descriptor.js]: remove isArray

File: lib/data/directory-descriptor.js [hunk 8]
```diff
@@ -16,6 +17,35 @@ class DirectoryDescriptor {
...
+      onTranslationsReady();
+      return;
+    }
+
+    const translationsIds = this.dataPackage.getTranslations().map(translation => translation.id);
```
Comment (rychkog) [inline lib/data/directory-descriptor.js]: _.map(this.dataPackage.getTranslations(), 'id')

File: lib/ddf-definitions/concept.js [hunk 9]
```diff
@@ -54,6 +56,30 @@ class Concept {
...
+
+  getTranslationsData(onTranslationsReady) {
+    const transActions = _.compact(_.flattenDeep(
+      this.fileDescriptors
+        .map(fileDescriptor =>
+          fileDescriptor.getExistingTranslationDescriptors()
+            .map(translationDesc => onTranslationLoaded => {
+              const transCollectionName = `translation-${translationDesc.file}`;
+
+              this.db
```
Comment (rychkog) [inline lib/ddf-definitions/concept.js]: this.db.fillCollection(transCollectionName, translationDesc.fullPath, err => {
        translationDesc.content = this.db.getCollection(transCollectionName);
        onTranslationLoaded(err);
 }, true);	
})

File: lib/ddf-definitions/ddf-data-set.js [hunk 10]
```diff
@@ -44,9 +44,12 @@ class DdfDataSet {
                     .fillCollection(
                       Symbol.keyFor(fileDescriptor.type),
                       fileDescriptor.fullPath,
-                      err => {
-                        this.expectedClass[fileDescriptor.type].addFileDescriptor(fileDescriptor);
-                        onFileLoaded(err);
+                      fileErr => {
+                        this.expectedClass[fileDescriptor.type]
```
Comment (rychkog) [inline lib/ddf-definitions/ddf-data-set.js]: one liners!

File: lib/ddf-rules/translation-rules/unexpected-translation-header.js [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+                if (transDesc.headers.length > headersDiff.length || headersDiff.length === 0) {
+                  issues.push(new Issue(registry.UNEXPECTED_TRANSLATION_HEADER)
+                    .setPath(transDesc.fullPath)
+                    .setData({
+                      reason: 'extra data in translation',
+                      ddfFileHeaders: fileDescriptor.headers,
+                      translationHeaders: transDesc.headers
+                    }));
+                }
+
```
Comment (rychkog) [inline lib/ddf-rules/translation-rules/unexpected-translation-header.js]: extract a function ?

File: index.js (hunk 4 above)
Comment (buchslava) [inline index.js]: I'll rename to 'issues'

**Pull request merged**